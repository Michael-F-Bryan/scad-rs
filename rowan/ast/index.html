<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Working with abstract syntax trees."><meta name="keywords" content="rust, rustlang, rust-lang, ast"><title>rowan::ast - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../../SourceSerif4-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../FiraSans-Regular.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../FiraSans-Medium.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../SourceCodePro-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../SourceSerif4-Bold.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../SourceCodePro-Semibold.ttf.woff2"><link rel="stylesheet" href="../../normalize.css"><link rel="stylesheet" href="../../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" href="../../ayu.css" disabled><link rel="stylesheet" href="../../dark.css" disabled><link rel="stylesheet" href="../../light.css" id="themeStyle"><script id="default-settings" ></script><script src="../../storage.js"></script><script defer src="../../main.js"></script><noscript><link rel="stylesheet" href="../../noscript.css"></noscript><link rel="alternate icon" type="image/png" href="../../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../../favicon-32x32.png"><link rel="icon" type="image/svg+xml" href="../../favicon.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="sidebar-logo" href="../../rowan/index.html"><div class="logo-container"><img class="rust-logo" src="../../rust-logo.svg" alt="logo"></div></a><h2 class="location"></h2></nav><nav class="sidebar"><a class="sidebar-logo" href="../../rowan/index.html"><div class="logo-container"><img class="rust-logo" src="../../rust-logo.svg" alt="logo"></div></a><h2 class="location"><a href="#">Module ast</a></h2><div class="sidebar-elems"><section><div class="block"><ul><li><a href="#modules">Modules</a></li><li><a href="#structs">Structs</a></li><li><a href="#traits">Traits</a></li></ul></div></section></div></nav><main><div class="width-limiter"><div class="sub-container"><a class="sub-logo-container" href="../../rowan/index.html"><img class="rust-logo" src="../../rust-logo.svg" alt="logo"></a><nav class="sub"><form class="search-form"><div class="search-container"><span></span><input class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><button type="button">?</button></div><div id="settings-menu" tabindex="-1"><a href="../../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../../wheel.svg"></a></div></div></form></nav></div><section id="main-content" class="content"><div class="main-heading"><h1 class="fqn"><span class="in-band">Module <a href="../index.html">rowan</a>::<wbr><a class="mod" href="#">ast</a><button id="copy-path" onclick="copy_path(this)" title="Copy item path to clipboard"><img src="../../clipboard.svg" width="19" height="18" alt="Copy item path"></button></span></h1><span class="out-of-band"><a class="srclink" href="../../src/rowan/ast.rs.html#1-206">source</a> · <a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span></div><details class="rustdoc-toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Working with abstract syntax trees.</p>
<p>In rowan, syntax trees are transient objects. That means that we create
trees when we need them, and tear them down to save memory. In this
architecture, hanging on to a particular syntax node for a long time is
ill-advisable, as that keeps the whole tree resident.</p>
<p>Instead, we provide a <a href="struct.SyntaxNodePtr.html" title="SyntaxNodePtr"><code>SyntaxNodePtr</code></a> type, which stores information about
the <em>location</em> of a particular syntax node in a tree. It’s a small type
which can be cheaply stored, and which can be resolved to a real
<a href="../api/struct.SyntaxNode.html" title="SyntaxNode"><code>SyntaxNode</code></a> when necessary.</p>
<p>We also provide an <a href="trait.AstNode.html" title="AstNode"><code>AstNode</code></a> trait for typed AST wrapper APIs over rowan
nodes.</p>
</div></details><h2 id="modules" class="small-section-header"><a href="#modules">Modules</a></h2><div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="mod" href="support/index.html" title="rowan::ast::support mod">support</a></div></div></div><h2 id="structs" class="small-section-header"><a href="#structs">Structs</a></h2><div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.AstChildren.html" title="rowan::ast::AstChildren struct">AstChildren</a></div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.AstPtr.html" title="rowan::ast::AstPtr struct">AstPtr</a></div><div class="item-right docblock-short"><p>Like <a href="struct.SyntaxNodePtr.html" title="SyntaxNodePtr"><code>SyntaxNodePtr</code></a>, but remembers the type of node.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.SyntaxNodePtr.html" title="rowan::ast::SyntaxNodePtr struct">SyntaxNodePtr</a></div><div class="item-right docblock-short"><p>A “pointer” to a <a href="../api/struct.SyntaxNode.html" title="SyntaxNode"><code>SyntaxNode</code></a>, via location in the source code.</p>
</div></div></div><h2 id="traits" class="small-section-header"><a href="#traits">Traits</a></h2><div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="trait" href="trait.AstNode.html" title="rowan::ast::AstNode trait">AstNode</a></div><div class="item-right docblock-short"><p>The main trait to go from untyped <a href="../api/struct.SyntaxNode.html" title="SyntaxNode"><code>SyntaxNode</code></a> to a typed AST. The
conversion itself has zero runtime cost: AST and syntax nodes have exactly
the same representation: a pointer to the tree root and a pointer to the
node itself.</p>
</div></div></div></section></div></main><div id="rustdoc-vars" data-root-path="../../" data-current-crate="rowan" data-themes="ayu,dark,light" data-resource-suffix="" data-rustdoc-version="1.65.0-nightly (b44197abb 2022-09-05)" ></div></body></html>