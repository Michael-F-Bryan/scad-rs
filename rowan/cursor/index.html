<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Implementation of the cursors – API for convenient access to syntax trees."><meta name="keywords" content="rust, rustlang, rust-lang, cursor"><title>rowan::cursor - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../../SourceSerif4-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../FiraSans-Regular.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../FiraSans-Medium.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../SourceCodePro-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../SourceSerif4-Bold.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../SourceCodePro-Semibold.ttf.woff2"><link rel="stylesheet" href="../../normalize.css"><link rel="stylesheet" href="../../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" href="../../ayu.css" disabled><link rel="stylesheet" href="../../dark.css" disabled><link rel="stylesheet" href="../../light.css" id="themeStyle"><script id="default-settings" ></script><script src="../../storage.js"></script><script defer src="../../main.js"></script><noscript><link rel="stylesheet" href="../../noscript.css"></noscript><link rel="alternate icon" type="image/png" href="../../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../../favicon-32x32.png"><link rel="icon" type="image/svg+xml" href="../../favicon.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="sidebar-logo" href="../../rowan/index.html"><div class="logo-container"><img class="rust-logo" src="../../rust-logo.svg" alt="logo"></div></a><h2 class="location"></h2></nav><nav class="sidebar"><a class="sidebar-logo" href="../../rowan/index.html"><div class="logo-container"><img class="rust-logo" src="../../rust-logo.svg" alt="logo"></div></a><h2 class="location"><a href="#">Module cursor</a></h2><div class="sidebar-elems"><section><div class="block"><ul><li><a href="#structs">Structs</a></li><li><a href="#types">Type Definitions</a></li></ul></div></section></div></nav><main><div class="width-limiter"><div class="sub-container"><a class="sub-logo-container" href="../../rowan/index.html"><img class="rust-logo" src="../../rust-logo.svg" alt="logo"></a><nav class="sub"><form class="search-form"><div class="search-container"><span></span><input class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><button type="button">?</button></div><div id="settings-menu" tabindex="-1"><a href="../../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../../wheel.svg"></a></div></div></form></nav></div><section id="main-content" class="content"><div class="main-heading"><h1 class="fqn"><span class="in-band">Module <a href="../index.html">rowan</a>::<wbr><a class="mod" href="#">cursor</a><button id="copy-path" onclick="copy_path(this)" title="Copy item path to clipboard"><img src="../../clipboard.svg" width="19" height="18" alt="Copy item path"></button></span></h1><span class="out-of-band"><a class="srclink" href="../../src/rowan/cursor.rs.html#1-1283">source</a> · <a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span></div><details class="rustdoc-toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Implementation of the cursors – API for convenient access to syntax trees.</p>
<p>Functional programmers will recognize that this module implements a zipper
for a purely functional (green) tree.</p>
<p>A cursor node (<code>SyntaxNode</code>) points to a <code>GreenNode</code> and a parent
<code>SyntaxNode</code>. This allows cursor to provide iteration over both ancestors
and descendants, as well as a cheep access to absolute offset of the node in
file.</p>
<p>By default <code>SyntaxNode</code>s are immutable, but you can get a mutable copy of
the tree by calling <code>clone_for_update</code>. Mutation is based on interior
mutability and doesn’t need <code>&amp;mut</code>. You can have two <code>SyntaxNode</code>s pointing
at different parts of the same tree; mutations via the first node will be
reflected in the other.</p>
</div></details><h2 id="structs" class="small-section-header"><a href="#structs">Structs</a></h2><div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.Preorder.html" title="rowan::cursor::Preorder struct">Preorder</a></div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.PreorderWithTokens.html" title="rowan::cursor::PreorderWithTokens struct">PreorderWithTokens</a></div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.SyntaxElementChildren.html" title="rowan::cursor::SyntaxElementChildren struct">SyntaxElementChildren</a></div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.SyntaxNode.html" title="rowan::cursor::SyntaxNode struct">SyntaxNode</a></div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.SyntaxNodeChildren.html" title="rowan::cursor::SyntaxNodeChildren struct">SyntaxNodeChildren</a></div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.SyntaxToken.html" title="rowan::cursor::SyntaxToken struct">SyntaxToken</a></div></div></div><h2 id="types" class="small-section-header"><a href="#types">Type Definitions</a></h2><div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="type" href="type.SyntaxElement.html" title="rowan::cursor::SyntaxElement type">SyntaxElement</a></div></div></div></section></div></main><div id="rustdoc-vars" data-root-path="../../" data-current-crate="rowan" data-themes="ayu,dark,light" data-resource-suffix="" data-rustdoc-version="1.65.0-nightly (b44197abb 2022-09-05)" ></div></body></html>