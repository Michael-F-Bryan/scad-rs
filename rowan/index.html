<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="A generic library for lossless syntax trees. See `examples/s_expressions.rs` for a tutorial."><meta name="keywords" content="rust, rustlang, rust-lang, rowan"><title>rowan - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Regular.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Medium.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Bold.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Semibold.ttf.woff2"><link rel="stylesheet" href="../normalize.css"><link rel="stylesheet" href="../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" href="../ayu.css" disabled><link rel="stylesheet" href="../dark.css" disabled><link rel="stylesheet" href="../light.css" id="themeStyle"><script id="default-settings" ></script><script src="../storage.js"></script><script defer src="../crates.js"></script><script defer src="../main.js"></script><noscript><link rel="stylesheet" href="../noscript.css"></noscript><link rel="alternate icon" type="image/png" href="../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../favicon-32x32.png"><link rel="icon" type="image/svg+xml" href="../favicon.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="sidebar-logo" href="../rowan/index.html"><div class="logo-container"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></div></a><h2 class="location"></h2></nav><nav class="sidebar"><a class="sidebar-logo" href="../rowan/index.html"><div class="logo-container"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></div></a><h2 class="location"><a href="#">Crate rowan</a></h2><div class="sidebar-elems"><div class="block"><ul><li class="version">Version 0.15.8</li><li><a id="all-types" href="all.html">All Items</a></li></ul></div><section><div class="block"><ul><li><a href="#reexports">Re-exports</a></li><li><a href="#modules">Modules</a></li><li><a href="#structs">Structs</a></li><li><a href="#enums">Enums</a></li><li><a href="#traits">Traits</a></li></ul></div></section></div></nav><main><div class="width-limiter"><div class="sub-container"><a class="sub-logo-container" href="../rowan/index.html"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></a><nav class="sub"><form class="search-form"><div class="search-container"><span></span><input class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><button type="button">?</button></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../wheel.svg"></a></div></div></form></nav></div><section id="main-content" class="content"><div class="main-heading"><h1 class="fqn"><span class="in-band">Crate <a class="mod" href="#">rowan</a><button id="copy-path" onclick="copy_path(this)" title="Copy item path to clipboard"><img src="../clipboard.svg" width="19" height="18" alt="Copy item path"></button></span></h1><span class="out-of-band"><a class="srclink" href="../src/rowan/lib.rs.html#1-41">source</a> · <a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span></div><details class="rustdoc-toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>A generic library for lossless syntax trees.
See <code>examples/s_expressions.rs</code> for a tutorial.</p>
</div></details><h2 id="reexports" class="small-section-header"><a href="#reexports">Re-exports</a></h2><div class="item-table"><div class="item-row"><div class="item-left import-item" id="reexport.Language"><code>pub use crate::api::<a class="trait" href="api/trait.Language.html" title="trait rowan::api::Language">Language</a>;</code></div></div><div class="item-row"><div class="item-left import-item" id="reexport.SyntaxElement"><code>pub use crate::api::<a class="type" href="api/type.SyntaxElement.html" title="type rowan::api::SyntaxElement">SyntaxElement</a>;</code></div></div><div class="item-row"><div class="item-left import-item" id="reexport.SyntaxElementChildren"><code>pub use crate::api::<a class="struct" href="api/struct.SyntaxElementChildren.html" title="struct rowan::api::SyntaxElementChildren">SyntaxElementChildren</a>;</code></div></div><div class="item-row"><div class="item-left import-item" id="reexport.SyntaxNode"><code>pub use crate::api::<a class="struct" href="api/struct.SyntaxNode.html" title="struct rowan::api::SyntaxNode">SyntaxNode</a>;</code></div></div><div class="item-row"><div class="item-left import-item" id="reexport.SyntaxNodeChildren"><code>pub use crate::api::<a class="struct" href="api/struct.SyntaxNodeChildren.html" title="struct rowan::api::SyntaxNodeChildren">SyntaxNodeChildren</a>;</code></div></div><div class="item-row"><div class="item-left import-item" id="reexport.SyntaxToken"><code>pub use crate::api::<a class="struct" href="api/struct.SyntaxToken.html" title="struct rowan::api::SyntaxToken">SyntaxToken</a>;</code></div></div></div><h2 id="modules" class="small-section-header"><a href="#modules">Modules</a></h2><div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="mod" href="api/index.html" title="rowan::api mod">api</a></div></div><div class="item-row"><div class="item-left module-item"><a class="mod" href="ast/index.html" title="rowan::ast mod">ast</a></div><div class="item-right docblock-short"><p>Working with abstract syntax trees.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="mod" href="cursor/index.html" title="rowan::cursor mod">cursor</a></div><div class="item-right docblock-short"><p>Implementation of the cursors – API for convenient access to syntax trees.</p>
</div></div></div><h2 id="structs" class="small-section-header"><a href="#structs">Structs</a></h2><div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.Checkpoint.html" title="rowan::Checkpoint struct">Checkpoint</a></div><div class="item-right docblock-short"><p>A checkpoint for maybe wrapping a node. See <code>GreenNodeBuilder::checkpoint</code> for details.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.Children.html" title="rowan::Children struct">Children</a></div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.GreenNode.html" title="rowan::GreenNode struct">GreenNode</a></div><div class="item-right docblock-short"><p>Internal node in the immutable tree.
It has other nodes and tokens as children.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.GreenNodeBuilder.html" title="rowan::GreenNodeBuilder struct">GreenNodeBuilder</a></div><div class="item-right docblock-short"><p>A builder for a green tree.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.GreenNodeData.html" title="rowan::GreenNodeData struct">GreenNodeData</a></div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.GreenToken.html" title="rowan::GreenToken struct">GreenToken</a></div><div class="item-right docblock-short"><p>Leaf node in the immutable tree.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.GreenTokenData.html" title="rowan::GreenTokenData struct">GreenTokenData</a></div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.NodeCache.html" title="rowan::NodeCache struct">NodeCache</a></div><div class="item-right docblock-short"><p>Interner for GreenTokens and GreenNodes</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.SyntaxKind.html" title="rowan::SyntaxKind struct">SyntaxKind</a></div><div class="item-right docblock-short"><p>SyntaxKind is a type tag for each token or node.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.SyntaxText.html" title="rowan::SyntaxText struct">SyntaxText</a></div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.TextRange.html" title="rowan::TextRange struct">TextRange</a></div><div class="item-right docblock-short"><p>A range in text, represented as a pair of <a href="struct.TextSize.html" title="TextSize"><code>TextSize</code></a>.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.TextSize.html" title="rowan::TextSize struct">TextSize</a></div><div class="item-right docblock-short"><p>A measure of text length. Also, equivalently, an index into text.</p>
</div></div></div><h2 id="enums" class="small-section-header"><a href="#enums">Enums</a></h2><div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.Direction.html" title="rowan::Direction enum">Direction</a></div></div><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.NodeOrToken.html" title="rowan::NodeOrToken enum">NodeOrToken</a></div></div><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.TokenAtOffset.html" title="rowan::TokenAtOffset enum">TokenAtOffset</a></div><div class="item-right docblock-short"><p>There might be zero, one or two leaves at a given offset.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.WalkEvent.html" title="rowan::WalkEvent enum">WalkEvent</a></div><div class="item-right docblock-short"><p><code>WalkEvent</code> describes tree walking process.</p>
</div></div></div><h2 id="traits" class="small-section-header"><a href="#traits">Traits</a></h2><div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="trait" href="trait.TextLen.html" title="rowan::TextLen trait">TextLen</a></div><div class="item-right docblock-short"><p>Primitives with a textual length that can be passed to <a href="struct.TextSize.html#method.of" title="TextSize::of"><code>TextSize::of</code></a>.</p>
</div></div></div></section></div></main><div id="rustdoc-vars" data-root-path="../" data-current-crate="rowan" data-themes="ayu,dark,light" data-resource-suffix="" data-rustdoc-version="1.65.0-nightly (b44197abb 2022-09-05)" ></div></body></html>