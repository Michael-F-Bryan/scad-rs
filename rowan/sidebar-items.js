window.SIDEBAR_ITEMS = {"enum":[["Direction",""],["NodeOrToken",""],["TokenAtOffset","There might be zero, one or two leaves at a given offset."],["WalkEvent","`WalkEvent` describes tree walking process."]],"mod":[["api",""],["ast","Working with abstract syntax trees."],["cursor","Implementation of the cursors â€“ API for convenient access to syntax trees."]],"struct":[["Checkpoint","A checkpoint for maybe wrapping a node. See `GreenNodeBuilder::checkpoint` for details."],["Children",""],["GreenNode","Internal node in the immutable tree. It has other nodes and tokens as children."],["GreenNodeBuilder","A builder for a green tree."],["GreenNodeData",""],["GreenToken","Leaf node in the immutable tree."],["GreenTokenData",""],["NodeCache","Interner for GreenTokens and GreenNodes"],["SyntaxKind","SyntaxKind is a type tag for each token or node."],["SyntaxText",""],["TextRange","A range in text, represented as a pair of [`TextSize`][struct@TextSize]."],["TextSize","A measure of text length. Also, equivalently, an index into text."]],"trait":[["TextLen","Primitives with a textual length that can be passed to [`TextSize::of`]."]]};