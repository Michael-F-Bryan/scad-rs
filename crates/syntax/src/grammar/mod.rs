mod expressions;
mod token_set;
mod top_level;

pub use self::token_set::TokenSet;
pub(crate) use self::top_level::*;

use crate::{ast::Package, parser::Parser, SyntaxKind};
use rowan::{ast::AstNode, TextRange};

pub fn parse<'a>(
    tokens: impl IntoIterator<Item = (SyntaxKind, &'a str)>,
) -> (Package, Vec<ParseError>) {
    let mut parser = Parser::new(tokens);
    package(&mut parser);
    let (node, errors) = parser.finish();

    (Package::cast(node).unwrap(), errors)
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ParseError {
    pub location: TextRange,
    pub msg: String,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_the_kitchen_sink() {
        let src = r#"
            include "foo/bar.scad"
            use "baz.scad"
            a = 42;
            b = (42);
            c = 1 + 1;
        "#;

        let (pkg, errors) = parse(crate::tokenize(src));

        eprintln!("{pkg:#?}");

        let statements: Vec<_> = pkg.statements().collect();
        assert_eq!(statements.len(), 5);

        let _include = statements[0].as_include().unwrap();
        let _use = statements[1].as_use().unwrap();
        let _a = statements[2].as_assignment().unwrap();
        let _b = statements[3].as_assignment().unwrap();

        let c = statements[4].as_assignment().unwrap();
        let _bin_expr = c.value().unwrap().as_bin_expr().unwrap();

        insta::assert_debug_snapshot!(pkg);
        insta::assert_debug_snapshot!(errors);
    }
}
