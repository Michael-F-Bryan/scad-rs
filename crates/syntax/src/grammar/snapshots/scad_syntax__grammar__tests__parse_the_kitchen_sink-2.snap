---
source: crates/syntax/src/grammar/mod.rs
expression: errors
---
[]
